far/vmenu.hpp

```diff
@@ -147,16 +146,80 @@ struct MenuItemEx: menu_item
 	wchar_t AutoHotkey{};
 	size_t AutoHotkeyPos{};
 	std::list<std::pair<int, int>> Annotations;
+
+	int SafeGetFirstAnnotation() const noexcept { return Annotations.empty() ? 0 : Annotations.front().first; }
 };
 
 struct menu_layout;
+enum class item_hscroll_policy;
 
 struct SortItemParam
 {
 	bool Reverse;
 	int Offset;
 };
 
+// Keeps track of the horizontal state of all items.
+// The tracking is best effort. See comments below.
+// Everything is relative to menu_layout::TextArea::first (Left edge).
+class vmenu_horizontal_tracker
+{
+	struct cookie
+	{
+		NONCOPYABLE(cookie);
+		explicit cookie(vmenu_horizontal_tracker* Tracker) noexcept: m_Tracker{ Tracker } {}
+		~cookie() noexcept { m_Tracker->m_IsBulkUpdate = false; }
```

Comment:
dtors are noexcept by default, do we need to state it explicitly?

Reply:
No, we don't. Copy-paste. Fixed.